File: modules/brk-client/tests/hash_prefix.js
import assert from "node:assert/strict";
import { BrkClient, addressPayloadHashPrefix } from "../index.js";

const vectors = [
  [Uint8Array.of(0x4e, 0x73), "58101afa51a1ecfd"],
  [Uint8Array.from({ length: 20 }, (_, i) => i), "c3327ecb8ae1ff23"],
  [Uint8Array.from({ length: 32 }, (_, i) => i), "c0186990f026b180"],
  [Uint8Array.from({ length: 65 }, (_, i) => i), "0d4b77027ae7d700"],
];

for (const [payload, hash] of vectors) {
  assert.equal(addressPayloadHashPrefix(payload, 16), hash);
  assert.equal(BrkClient.addressPayloadHashPrefix(payload, 8), hash.slice(0, 8));
}

assert.throws(() => addressPayloadHashPrefix(Uint8Array.of(), 16), /non-empty/);
assert.throws(() => addressPayloadHashPrefix(new Uint8Array(66), 16), /at most 65/);
assert.throws(() => addressPayloadHashPrefix(Uint8Array.of(1, 2), 0), /1 to 16/);

const client = new BrkClient("http://127.0.0.1:0");
client.getAddressHashPrefixMatches = (addrType, prefix) => ({ addrType, prefix, truncated: false, addresses: [] });

assert.deepEqual(
  client.getAddressPayloadHashPrefixMatches("p2pkh", Uint8Array.from({ length: 20 }, (_, i) => i), 8),
  { addrType: "p2pkh", prefix: "c3327ecb", truncated: false, addresses: [] },
);
for (const [addrType, length] of [["p2pk33", 33], ["p2pk65", 65]]) {
  const payload = Uint8Array.from({ length }, (_, i) => i);
  assert.deepEqual(
    client.getAddressPayloadHashPrefixMatches(addrType, payload, 8),
    {
      addrType,
      prefix: addressPayloadHashPrefix(payload, 8),
      truncated: false,
      addresses: [],
    },
  );
}
assert.throws(
  () => client.getAddressPayloadHashPrefixMatches("p2pkh", Uint8Array.of(1, 2), 8),
  /p2pkh address payload length 20 bytes/,
);
assert.throws(
  () => client.getAddressPayloadHashPrefixMatches("p2pk33", new Uint8Array(65), 8),
  /p2pk33 address payload length 33 bytes/,
);
assert.throws(
  () => client.getAddressPayloadHashPrefixMatches("p2pk65", new Uint8Array(33), 8),
  /p2pk65 address payload length 65 bytes/,
);
assert.throws(
  () => client.getAddressPayloadHashPrefixMatches("p2pk", new Uint8Array(33), 8),
  /Unsupported address type/,
);
assert.throws(
  () => client.getAddressPayloadHashPrefixMatches("opreturn", Uint8Array.of(1, 2), 8),
  /Unsupported address type/,
);

File: crates/brk_client/tests/hash_prefix.rs
use brk_client::{
    BrkClient, OutputType, address_hash_prefix, address_payload_hash_prefix, decode_address_payload,
};

#[test]
fn address_payload_hash_prefix_vectors() {
    let vectors = [
        (vec![0x4e, 0x73], "58101afa51a1ecfd"),
        ((0_u8..20).collect::<Vec<_>>(), "c3327ecb8ae1ff23"),
        ((0_u8..32).collect::<Vec<_>>(), "c0186990f026b180"),
        ((0_u8..65).collect::<Vec<_>>(), "0d4b77027ae7d700"),
    ];

    for (payload, expected) in vectors.iter() {
        assert_eq!(address_payload_hash_prefix(payload, 16).unwrap(), *expected);
        assert_eq!(
            BrkClient::address_payload_hash_prefix(payload, 8).unwrap(),
            &expected[..8]
        );
    }
}

#[test]
fn address_payload_hash_prefix_validation() {
    assert!(address_payload_hash_prefix(&[], 16).is_err());
    assert!(address_payload_hash_prefix(&[0; 66], 16).is_err());
    assert!(address_payload_hash_prefix(&[1, 2], 0).is_err());
    assert!(address_payload_hash_prefix(&[1, 2], 17).is_err());
}

#[test]
fn address_hash_prefix_uses_brk_address_parser() {
    let address = "1BoatSLRHtKNngkdXEeobR76b53LETtpyT";
    let decoded = decode_address_payload(address).unwrap();
    assert_eq!(decoded.addr_type, OutputType::P2PKH);
    assert_eq!(decoded.payload.len(), 20);

    let hashed = address_hash_prefix(address, 8).unwrap();
    assert_eq!(hashed.addr_type, OutputType::P2PKH);
    assert_eq!(
        hashed.prefix,
        address_payload_hash_prefix(&decoded.payload, 8).unwrap()
    );
}

#[test]
fn p2pk_payload_types_are_distinct() {
    let p2pk33 = decode_address_payload(
        "0279be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798",
    )
    .unwrap();
    assert_eq!(p2pk33.addr_type, OutputType::P2PK33);
    assert_eq!(p2pk33.payload.len(), 33);

    let p2pk65 = decode_address_payload(
        "04678afdb0fe5548271967f1a67130b7105cd6a828e03909a67962e0ea1f61deb649f6bc3f4cef38c4f35504e51ec112de5c384df7ba0b8d578a4c702b6bf11d5f",
    )
    .unwrap();
    assert_eq!(p2pk65.addr_type, OutputType::P2PK65);
    assert_eq!(p2pk65.payload.len(), 65);
}
